src/GafferImage/ImageWriter.cpp

```diff
@@ -1214,26 +1131,126 @@ void setImageSpecFormatOptions( const ImageWriter *node, ImageSpec *spec, const
+	}
+	else if( dataType == "int16" )
+	{
+		return TypeDesc::INT16;
+	}
+	else if( dataType == "int32" )
+	{
+		return TypeDesc::INT32;
+	}
+	else if( dataType == "int64" )
+	{
+		return TypeDesc::INT64;
+	}
+	else if( dataType == "uint8" )
+	{
+		return TypeDesc::UINT8;
+	}
+	else if( dataType == "uint16" )
+	{
+		return TypeDesc::UINT16;
+	}
+	else if( dataType == "uint32" )
+	{
+		return TypeDesc::UINT32;
+	}
+	else if( dataType == "uint64" )
+	{
+		return TypeDesc::UINT64;
+	}
+	else if( dataType == "half" )
+	{
+		return TypeDesc::HALF;
+	}
+	else if( dataType == "float" )
+	{
+		return TypeDesc::FLOAT;
+	}
+	else if( dataType == "double" )
+	{
+		return TypeDesc::DOUBLE;
+	}
+	else if( fileFormatName == "dpx" && dataType == "uint10" )
+	{
+		if( !spec )
 		{
-			spec->set_format( TypeDesc::UINT16 );
-			spec->attribute ("oiio:BitsPerSample", 10);
+			// In order to add this to the spec, we need to be processing the datatype for the whole image at
+			// once, in which case we'll get passed the spec.
+			throw IECore::Exception( "Cannot use data type uint10 unless all channels have matching dataType" );
 		}
-		else if( dataType == "uint12" )
+		spec->attribute ("oiio:BitsPerSample", 10);
+		return TypeDesc::UINT16;
+	}
+	else if( fileFormatName == "dpx" && dataType == "uint12" )
+	{
+		if( !spec )
 		{
-			spec->set_format( TypeDesc::UINT16 );
-			spec->attribute ("oiio:BitsPerSample", 12);
+			// In order to add this to the spec, we need to be processing the datatype for the whole image at
+			// once, in which case we'll get passed the spec.
+			throw IECore::Exception( "Cannot use data type uint12 unless all channels have matching dataType" );
 		}
+		spec->attribute ("oiio:BitsPerSample", 12);
+		return TypeDesc::UINT16;
 	}
-	else if( fileFormatName == "png" )
+
+
+	throw IECore::Exception( "Unknown data type for " + fileFormatName + " file write : " + dataType );
+}
+
+void setImageSpecFormatOptions( const ImageWriter *node, const ImageOutput *out, ImageSpec *spec, const std::vector<std::string> &channelDataTypes, const std::string &fileFormatName )
+{
+	if( !channelDataTypes.size() )
 	{
-		spec->attribute( "png:compressionLevel", optionsPlug->getChild<IntPlug>( g_compressionLevelPlugName )->getValue() );
+		return;
 	}
-	else if( fileFormatName == "webp" )
+
+	bool allSame = std::all_of( channelDataTypes.begin(), channelDataTypes.end(),
+		[channelDataTypes] ( const std::string& t ) { return t == channelDataTypes.front(); }
+	);
+
+	if( allSame )
 	{
-		spec->attribute( "CompressionQuality", optionsPlug->getChild<IntPlug>( g_compressionQualityPlugName )->getValue() );
+		if( channelDataTypes.front() == "" )
+		{
+			return;  // Nothing is setting a data type
+		}
+
+		// Since all the channels have the same data type, it's OK if the data type affects
+		// other parts of the spec, so we pass the spec in for possible modification
+		spec->set_format( dataTypeToOiioType( channelDataTypes[0], fileFormatName, spec ) );
```

Comment:
I found this a bit confusing - `dataTypeToOiioType` doesn't _sound_ like it's going to have side effects. It seems like it'd be pretty straightforward to do the DPX special case here instead, unless I missed something?